test/unit/auth/auth.spec.ts

```diff
@@ -200,2602 +219,2852 @@ function getSAMLConfigServerResponse(providerId: string): SAMLConfigServerRespon
-        .should.eventually.be.fulfilled;
-    });
+      describe('app', () => {
+        it('returns the app from the constructor', () => {
+          // We expect referential equality here
+          expect((auth as Auth).app).to.equal(mockApp);
+        });
 
-    it('should be fulfilled given an app which returns invalid access tokens', () => {
-      // createCustomToken() does not rely on an access token and therefore works in this scenario.
-      return malformedAccessTokenAuth.createCustomToken(mocks.uid, mocks.developerClaims)
-        .should.eventually.be.fulfilled;
-    });
+        it('is read-only', () => {
+          expect(() => {
+            (auth as any).app = mockApp;
+          }).to.throw('Cannot set property app of #<Auth> which has only a getter');
+        });
+      });
 
-    it('should be fulfilled given an app which fails to generate access tokens', () => {
-      // createCustomToken() does not rely on an access token and therefore works in this scenario.
-      return rejectedPromiseAccessTokenAuth.createCustomToken(mocks.uid, mocks.developerClaims)
-        .should.eventually.be.fulfilled;
-    });
-  });
+      describe('Auth.forTenant()', () => {
+        it('should cache the returned TenantAwareAuth', () => {
+          const tenantAwareAuth1 = (auth as Auth).forTenant('tenantId1');
+          const tenantAwareAuth2 = (auth as Auth).forTenant('tenantId2');
+          expect((auth as Auth).forTenant('tenantId1')).to.equal(tenantAwareAuth1);
+          expect((auth as Auth).forTenant('tenantId2')).to.equal(tenantAwareAuth2);
+          expect(tenantAwareAuth1).to.not.be.equal(tenantAwareAuth2);
+          expect(tenantAwareAuth1.tenantId).to.equal('tenantId1');
+          expect(tenantAwareAuth2.tenantId).to.equal('tenantId2');
+        });
+      });
+    } else {
+      // Run tests for TenantAwareAuth.
+      describe('forTenant()', () => {
+        const invalidTenantIds = [null, NaN, 0, 1, true, false, '', [], [1, 'a'], {}, { a: 1 }, _.noop];
+        invalidTenantIds.forEach((invalidTenantId) => {
+          it('should throw given invalid tenant ID: ' + JSON.stringify(invalidTenantId), () => {
+            expect(() => {
+              const testAuth = new Auth(mockApp);
+              return testAuth.forTenant(invalidTenantId as any);
+            }).to.throw('The tenant ID must be a valid non-empty string.');
+          });
+        });
 
-  it('verifyIdToken() should throw when project ID is not specified', () => {
-    const mockCredentialAuth = new Auth(mocks.mockCredentialApp());
-    const expected = 'Must initialize app with a cert credential or set your Firebase project ID ' +
-      'as the GOOGLE_CLOUD_PROJECT environment variable to call verifyIdToken().';
-    expect(() => {
-      mockCredentialAuth.verifyIdToken(mocks.generateIdToken());
-    }).to.throw(expected);
-  });
+        it('should not throw on valid tenant ID', () => {
```

Comment:
This is pretty much the same as the following test case. Consider removing it.

Reply:
done